src/ZEO/asyncio/client.py

```diff
@@ -657,7 +655,26 @@ def tpc_finish_threadsafe(self, future, wait_ready, tid, updates, f):
                 self.protocol.close()
                 self.disconnected(self.protocol)
             else:
-                f(tid)
+                if getattr(f, "invalidateTransaction", False):
+                    # ``f`` is set up to invalidate the transaction
+                    # This is the case for "normal" operations involving
+                    # the ``MVCCAdapter``.
+                    # From ``ZODB>=5.6`` on, ``MVCCAdapter`` requires
+                    # that `lastTransaction` changes after invalidation
+                    # processing; therefore, `setLastTid` is called
+                    # after the callback
+                    f(tid)
+                    cache.setLastTid(tid)
+                else:  # called from tests
+                    # the call below would deadlock for older ``ZODB``
```

Comment:
@d-maurer, while I appreciate your work on getting ZEO5 fixed, to me going via this ad-hoc way would not be good. Like @jmuchemb said ZODB < 5.6 is already at risk of data corruption (see https://github.com/zopefoundation/ZODB/issues/290). From this point of view maybe more logical would be to just make ZEO5 require ZODB>=5.6 ?

I have also left some comments regarding `check_tid_ordering_w_commit` logic in https://github.com/zopefoundation/ZODB/pull/307#issuecomment-818926431 .